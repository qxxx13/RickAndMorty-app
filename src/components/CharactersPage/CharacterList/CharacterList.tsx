import React, { useContext } from "react";
import { useMemo, useEffect, useCallback } from "react";
import { useQuery } from "@apollo/client";

import { GET_CHARACTERS } from "../../../apollo/CharactersAPI";
import { CircularProgress, Container, Pagination, Stack } from "@mui/material";
import { CharacterListItem } from "./CharacterListItem/CharacterListItem";
import { CharacterItemModel } from "../../../models/CharacterItemModel";
import { SearchField } from "../SearchField/SearchField";
import { CharactersModel } from "../../../models/CharactersModel";
import { AppContext } from "../../../context/AppContext";
import { ActionType } from "../../../context/Actions";
import { GET_SEARCH_VALUE } from "../../../apollo/variables";
import { GetSearchValueModel } from "../../../models/VariablesModel";

export const CharacterList: React.FC = () => {
    const { state, dispatch } = useContext(AppContext);

    const { data: searchValueData } = useQuery<GetSearchValueModel>(GET_SEARCH_VALUE);

    const { loading, data } = useQuery<CharactersModel>(GET_CHARACTERS, {
        variables: {
            page: state.currentPage,
            filter: {
                name: searchValueData?.searchValue,
            },
        },
    });

    const onPaginationChange = useCallback(
        (event: React.ChangeEvent<unknown>, page: number) => {
            dispatch({ type: ActionType.SET_CURRENT_PAGE, payload: page });
        },
        [dispatch]
    );

    const CharacterList = useMemo(() => {
        if (!loading)
            return data?.characters.results.map((character: CharacterItemModel) => (
                <CharacterListItem character={character} key={character.id} />
            ));
    }, [data, loading]);

    useEffect(() => {
        localStorage.setItem("currentPage", String(state.currentPage));
    }, [state.currentPage]);

    return (
        <Container>
            <SearchField />
            <Stack justifyContent="center" flexDirection="row" sx={{ padding: "16px 0 16px 0" }}>
                <Pagination
                    count={data?.characters.info.pages}
                    onChange={onPaginationChange}
                    page={state.currentPage}
                />
            </Stack>
            <Stack alignItems="center">{!loading ? CharacterList : <CircularProgress />}</Stack>
        </Container>
    );
};
