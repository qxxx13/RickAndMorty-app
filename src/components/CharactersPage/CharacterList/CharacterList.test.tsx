import { render, screen } from "@testing-library/react";
import { CharacterList } from "./CharacterList";

const data = {
    characters: {
        info: {
            count: 1,
            pages: 1,
        },
        results: [
            {
                id: "1",
                name: "Test",
                image: "string",
                gender: "string",
                status: "string",
                species: "string",
                episode: [
                    {
                        id: "string",
                        name: "string",
                    },
                ],
            },
        ],
    },
};

describe("CharacterList component", () => {
    it("CharacterList render", () => {
        render(<CharacterList loading={false} data={data} />);

        expect(screen.getByRole("img")).toBeInTheDocument();
        expect(screen.getByText("Test")).toBeInTheDocument();
    });

    it("CharacterList render without data", () => {
        render(<CharacterList loading={false} />);

        expect(screen.queryByRole("img")).toBeNull();
    });

    it("CharacterList snapshot", () => {
        const view = render(<CharacterList loading={false} data={data} />);

        expect(view).toMatchSnapshot();
    });

    it("CharacterList epmty snapshot", () => {
        const view = render(<CharacterList loading={false} />);

        expect(view).toMatchSnapshot();
    });
});
