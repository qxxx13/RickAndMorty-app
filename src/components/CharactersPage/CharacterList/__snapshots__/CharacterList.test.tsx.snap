// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterList component CharacterList epmty snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiStack-root css-163ljxj-MuiStack-root"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiStack-root css-163ljxj-MuiStack-root"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CharacterList component CharacterList snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiStack-root css-163ljxj-MuiStack-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-14id3zl-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardMedia-root css-3hlzln-MuiCardMedia-root"
            role="img"
            style="background-image: url(string);"
          />
          <div
            class="MuiCardContent-root css-xrjuk6-MuiCardContent-root"
          >
            <h3
              class="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
            >
              Test
            </h3>
            <p
              class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            >
              string
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            >
              string
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            >
              string
            </p>
          </div>
          <div
            class="MuiStack-root css-shswq-MuiStack-root"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5 css-1jcm1rz-MuiTypography-root"
            >
              Episode:
            </h5>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              string
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiStack-root css-163ljxj-MuiStack-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-14id3zl-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardMedia-root css-3hlzln-MuiCardMedia-root"
          role="img"
          style="background-image: url(string);"
        />
        <div
          class="MuiCardContent-root css-xrjuk6-MuiCardContent-root"
        >
          <h3
            class="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
          >
            Test
          </h3>
          <p
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            string
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            string
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            string
          </p>
        </div>
        <div
          class="MuiStack-root css-shswq-MuiStack-root"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 css-1jcm1rz-MuiTypography-root"
          >
            Episode:
          </h5>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            string
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
